File: tl/src/main/java/com/github/badoualy/telegram/tl/api/requests/TLRequestInitConnection.java

package com.github.badoualy.telegram.tl.api.requests;

import com.github.badoualy.telegram.tl.TLContext;
import com.github.badoualy.telegram.tl.core.TLMethod;
import com.github.badoualy.telegram.tl.core.TLObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static com.github.badoualy.telegram.tl.StreamUtils.readInt;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLMethod;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeInt;
import static com.github.badoualy.telegram.tl.StreamUtils.writeTLMethod;
import static com.github.badoualy.telegram.tl.StreamUtils.writeTLString;


public class TLRequestInitConnection<T extends TLObject> extends TLMethod<T> {

    public static final int CLASS_ID = 0x69796de9;

    public int getClassId() {
        return CLASS_ID;
    }


    public TLRequestInitConnection(        int _apiId,         String _deviceModel,         String _systemVersion,         String _appVersion,         String _langCode,         TLMethod<T> _query) {
        this.apiId = _apiId;
        this.deviceModel = _deviceModel;
        this.systemVersion = _systemVersion;
        this.appVersion = _appVersion;
        this.langCode = _langCode;
        this.query = _query;

    }



    public T deserializeResponse(InputStream stream, TLContext context) throws IOException {

        return (T) query.deserializeResponse(stream, context);

    }
        


    protected int apiId;

    protected String deviceModel;

    protected String systemVersion;

    protected String appVersion;

    protected String langCode;

    protected TLMethod<T> query;


    public int getApiId() {
        return apiId;
    }

    public void setApiId(int value) {
        this.apiId = value;
    }

    public String getDeviceModel() {
        return deviceModel;
    }

    public void setDeviceModel(String value) {
        this.deviceModel = value;
    }

    public String getSystemVersion() {
        return systemVersion;
    }

    public void setSystemVersion(String value) {
        this.systemVersion = value;
    }

    public String getAppVersion() {
        return appVersion;
    }

    public void setAppVersion(String value) {
        this.appVersion = value;
    }

    public String getLangCode() {
        return langCode;
    }

    public void setLangCode(String value) {
        this.langCode = value;
    }

    public TLMethod<T> getQuery() {
        return query;
    }

    public void setQuery(TLMethod<T> value) {
        this.query = value;
    }


    @Override
    public void serializeBody(OutputStream stream) throws IOException {

        writeInt(this.apiId, stream);
        writeTLString(this.deviceModel, stream);
        writeTLString(this.systemVersion, stream);
        writeTLString(this.appVersion, stream);
        writeTLString(this.langCode, stream);
        writeTLMethod(this.query, stream);
    }


    @Override
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {

        this.apiId = readInt(stream);
        this.deviceModel = readTLString(stream);
        this.systemVersion = readTLString(stream);
        this.appVersion = readTLString(stream);
        this.langCode = readTLString(stream);
        this.query = readTLMethod(stream, context);
    }



    @Override
    public String toString() {
        return "initConnection#69796de9";
    }

}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/api/request/TLRequestInitConnection.kt
package com.github.badoualy.telegram.tl.api.request

import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_BOOLEAN
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_DOUBLE
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLBytesSerializedSize
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize
import com.github.badoualy.telegram.tl.api.TLInputClientProxy
import com.github.badoualy.telegram.tl.core.TLMethod
import com.github.badoualy.telegram.tl.core.TLObject
import com.github.badoualy.telegram.tl.serialization.TLDeserializer
import com.github.badoualy.telegram.tl.serialization.TLSerializer
import java.io.IOException
import kotlin.Any
import kotlin.Boolean
import kotlin.Int
import kotlin.String
import kotlin.jvm.Throws

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
class TLRequestInitConnection<T : TLObject>() : TLMethod<T>() {
    var apiId: Int = 0

    var deviceModel: String = ""

    var systemVersion: String = ""

    var appVersion: String = ""

    var systemLangCode: String = ""

    var langPack: String = ""

    var langCode: String = ""

    var proxy: TLInputClientProxy? = null

    var query: TLMethod<T>? = null

    private val _constructor: String = "initConnection#785188b8"

    override val constructorId: Int = CONSTRUCTOR_ID

    constructor(
            apiId: Int,
            deviceModel: String,
            systemVersion: String,
            appVersion: String,
            systemLangCode: String,
            langPack: String,
            langCode: String,
            proxy: TLInputClientProxy?,
            query: TLMethod<T>?
    ) : this() {
        this.apiId = apiId
        this.deviceModel = deviceModel
        this.systemVersion = systemVersion
        this.appVersion = appVersion
        this.systemLangCode = systemLangCode
        this.langPack = langPack
        this.langCode = langCode
        this.proxy = proxy
        this.query = query
    }

    @Throws(IOException::class)
    override fun deserializeResponse_(tlDeserializer: TLDeserializer): T = query!!.deserializeResponse(tlDeserializer)

    protected override fun computeFlags() {
        _flags = 0
        updateFlags(proxy, 1)
    }

    @Throws(IOException::class)
    override fun serializeBody(tlSerializer: TLSerializer) = with (tlSerializer)  {
        computeFlags()

        writeInt(_flags)
        writeInt(apiId)
        writeString(deviceModel)
        writeString(systemVersion)
        writeString(appVersion)
        writeString(systemLangCode)
        writeString(langPack)
        writeString(langCode)
        doIfMask(proxy, 1) { writeTLObject(it) }
        writeTLMethod(query!!)
    }

    @Throws(IOException::class)
    override fun deserializeBody(tlDeserializer: TLDeserializer) = with (tlDeserializer)  {
        _flags = readInt()
        apiId = readInt()
        deviceModel = readString()
        systemVersion = readString()
        appVersion = readString()
        systemLangCode = readString()
        langPack = readString()
        langCode = readString()
        proxy = readIfMask(1) { readTLObject<TLInputClientProxy>(TLInputClientProxy::class, TLInputClientProxy.CONSTRUCTOR_ID) }
        query = readTLMethod()
    }

    override fun computeSerializedSize(): Int {
        computeFlags()

        var size = SIZE_CONSTRUCTOR_ID
        size += SIZE_INT32
        size += SIZE_INT32
        size += computeTLStringSerializedSize(deviceModel)
        size += computeTLStringSerializedSize(systemVersion)
        size += computeTLStringSerializedSize(appVersion)
        size += computeTLStringSerializedSize(systemLangCode)
        size += computeTLStringSerializedSize(langPack)
        size += computeTLStringSerializedSize(langCode)
        size += getIntIfMask(proxy, 1) { it.computeSerializedSize() }
        size += query!!.computeSerializedSize()
        return size
    }

    override fun toString() = _constructor

    override fun equals(other: Any?): Boolean {
        if (other !is TLRequestInitConnection<*>) return false
        if (other === this) return true

        return _flags == other._flags
                && apiId == other.apiId
                && deviceModel == other.deviceModel
                && systemVersion == other.systemVersion
                && appVersion == other.appVersion
                && systemLangCode == other.systemLangCode
                && langPack == other.langPack
                && langCode == other.langCode
                && proxy == other.proxy
                && query == other.query
    }
    companion object  {
        const val CONSTRUCTOR_ID: Int = 0x785188b8.toInt()
    }
}
